### PR TITLE
PersistentEntity to glue together Sharding and PersistentBehavior better

### PR DESCRIPTION
**Pull request by patriknw**
* for example you don't have to worry about the persistenceId, only EntityTypeKey and entityId

PersistentEntityCompileOnlyTest illustrates how it's used.

### CODE REVIEW

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/PersistentEntity.scala [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+}
+
+object PersistentEntityRegistry {
+  def get(system: ActorSystem[_]): PersistentEntityRegistry =
+    new PersistentEntityRegistry(ClusterSharding.get(system))
+
+}
+
+class PersistentEntityRegistry(sharding: ClusterSharding) {
```
Comment (johanandren) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/PersistentEntity.scala]: I think bringing the whole concept of registry in is a mistake. Can we instead identify the pain points and make sharding easier to use for them?

Comment (patriknw) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/PersistentEntity.scala]: one concrete pain point is `persistenceId` vs `entityId`, there is no need to care about the `persistenceId` when using sharding, since that should be derived from `entityId`

**Comment by patriknw** [comment]
This PR now illustrates 3 alternatives:

1. status quo, no specific PersistentEntity, using ordinary ClusterSharding.start(ShardedEntity(entityId=>Behavior)), see PersistentEntity0CompileOnlyTest.java
1. PeristentEntity and PersistentEntityRegistry, see PersistentEntityCompileOnlyTest.java
2. PersistentEntity and new start method in ClusterSharding, no PersistentEntityRegistry, see PersistentEntity2CompileOnlyTest.java

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala [hunk 2]
```diff
@@ -231,6 +243,58 @@ object ShardedEntity {
...
+   * settings can be defined using the with` methods of the returned [[ShardedEntity]].
+   *
+   * @param createBehavior Create the behavior for an entity given an entityId
+   * @param typeKey A key that uniquely identifies the type of entity in this cluster
+   * @param stopMessage Message sent to an entity to tell it to stop, e.g. when rebalanced or passivated.
+   *
+   * @tparam Command The type of message the entity accepts
+   */
+  def ofPersistentEntity[Command, Event, State >: Null](
+    createPersistentEntity: JFunction[String, PersistentEntity2[Command, Event, State]],
```
Comment (patriknw) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: I did a small refinement here. Instead of having two overloaded `start` methods I added this factory method for `ShardedEntity`. See usage in [PersistentEntity3CompileOnlyTest](https://github.com/akka/akka/pull/25697/files#diff-0b7d0e19bcd9431650320ad44d241a74).

This should be equally discoverable and then we don't have to have a different two variants of ShardedEntity that are exactly the same aside from this factory.

Comment (viktorklang) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: Perhaps `ofPersistent` since the non-persistent one is merely called `of`?

Comment (jboner) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: I like it. But, actually, I was rather thinking that `ofEntity` and of `ofPersistentEntity` is more clear and guided.

Comment (WadeWaldron) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: I think I would go with "entityOf" and "persistentEntityOf".

Comment (patriknw) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: We should read this as:
```
ShardedEntity.of(...)
ShardedEntity.ofPersistentEntity(...)
```

I like @viktorklang 's proposal:
```
ShardedEntity.of(...)
ShardedEntity.ofPersistent(...)
```

@jonas 's suggestion also works for me although it's a bit of repetition.

another one could be:
```
ShardedEntity.ofBehavior(...)
ShardedEntity.ofPersistentEntity(...)
```

@WadeWaldron 's :
```
ShardedEntity.entityOf(...)
ShardedEntity.persistentEntityOf(...)
```
while that might read nice once it's written it's more difficult to see that these are two alternatives to choose between when using code completion in IDE. I like starting such with same prefix. Another reason for using `of` is that it's a pretty well know naming for factor methods in Java.

PS. Note that the `.of` method also accepts PersistentEntity since PersistentEntity __is a__ Behavior. The only purpose for having both is to give the user a better hint of the existence of PersistentEntity, which was what you asked for as I understand it.

Comment (viktorklang) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: I think they need to start with the same word so that they are adjacent in an auto-complete/intellisense situation. Therefor I prefer "of" and "ofPersistent"

Comment (jboner) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: I agree that starting with `of` is important.

Comment (renatocaval) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: Also agree with staring with `of` for discoverability, but also because `..Entity.enity..` is kind of repetitive. 

For alignment, we should use either the term `Entity` or `Behavior`, but not mixing them. 

```scala
ShardedEntity.ofBehavior(...)
ShardedEntity.ofPersistentBehavior(...)
```

```scala
ShardedEntity.ofEntity(...)
ShardedEntity.ofPersistentEntity(...)
```

I personally think that it should be `Behavior`. We are creating an `EntityRef` from a `Behavior`.

Comment (WadeWaldron) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: Fair argument. I can go with `of...` for discoverability.

**Comment by patriknw** [comment]
I have updated this. Still have to add some tests and update reference docs but now is a good time for pre-final review.

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala [hunk 3]
```diff
@@ -201,43 +202,56 @@ object ShardedEntity {
...
+   * @tparam Command The type of message the entity accepts
    */
-  def create[M](
-    createBehavior: BiFunction[ActorRef[ClusterSharding.ShardCommand], String, Behavior[M]],
-    typeKey:        EntityTypeKey[M],
-    stopMessage:    M): ShardedEntity[M, ShardingEnvelope[M]] =
-    new ShardedEntity(createBehavior, typeKey, stopMessage, Props.empty, Optional.empty(), Optional.empty(), Optional.empty())
+  def ofPersistentEntity[Command, Event, State >: Null](
+    typeKey:                EntityTypeKey[Command],
+    createPersistentEntity: JFunction[ShardedEntityContext[Command], PersistentEntity[Command, Event, State]],
```
Comment (johanandren) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: Should we avoid type rename/aliases like this in public APIs or do they look alright from Java?


Also the null bound is weird, but I think I saw another issue mentioning looking into that?

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala [hunk 4]
```diff
@@ -276,18 +290,26 @@ final class ShardedEntity[M, E] private[akka] (
...
 
 }
 
+/**
+ * Parameter to [[ShardedEntity.of]]
+ */
+final class ShardedEntityContext[M](
+  val entityId:     String,
+  val shard:        ActorRef[ClusterSharding.ShardCommand],
+  val actorContext: ActorContext[M])
```
Comment (johanandren) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: Java-style getters?

File: akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    @Override
+    public EventHandler<KnownPeople, Greeted> eventHandler() {
+      return this::applyEvent;
+    }
+
+    private KnownPeople applyEvent(KnownPeople state, Greeted evt) {
+      return state.add(evt.whom);
+    }
```
Comment (johanandren) [inline akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java]: I found this one a bit harder to read than explicitly saying in `(state, evt) -> state.add(evt)` in `eventHandler` (because of the extra indirection with no obvious gain perhaps)

File: akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  val HelloWorldEntityTypeKey: EntityTypeKey[HelloWorldCommand] =
+    EntityTypeKey[HelloWorldCommand]("HelloWorld")
+
+  def helloWorldBehavior(entityId: String): Behavior[HelloWorldCommand] = PersistentEntity(
+    entityTypeKey = HelloWorldEntityTypeKey,
+    entityId = entityId,
+    emptyState = KnownPeople(Set.empty),
+    commandHandler,
+    eventHandler
+  )
```
Comment (johanandren) [inline akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.scala]: Maybe an object as wrapping context for the persistent behavior would make it nice and easier to read?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/PersistentBehavior.scala [hunk 7]
```diff
@@ -70,7 +71,10 @@ object PersistentBehavior {
 
 }
 
-trait PersistentBehavior[Command, Event, State] extends DeferredBehavior[Command] {
+@DoNotInherit trait PersistentBehavior[Command, Event, State] extends DeferredBehavior[Command] {
```
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/PersistentBehavior.scala]: `/** Not for user extension */`

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala [hunk 8]
```diff
@@ -18,7 +18,7 @@ import scala.util.{ Failure, Success }
 
 /** Java API */
 @ApiMayChange
-abstract class PersistentBehavior[Command, Event, State >: Null] private (val persistenceId: PersistenceId, supervisorStrategy: Option[BackoffSupervisorStrategy]) extends DeferredBehavior[Command] {
+abstract class PersistentBehavior[Command, Event, State >: Null] private[akka] (val persistenceId: PersistenceId, supervisorStrategy: Option[BackoffSupervisorStrategy]) extends DeferredBehavior[Command] {
```
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala]: Does this constructor really become private from Java?

**Review by johanandren — COMMENTED** [review]
Looks good in general, added some comments

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala (hunk 4 above)
Comment (patriknw) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: ok

File: akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java (hunk 5 above)
Comment (patriknw) [reply akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java]: ok, perhaps overused method references

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala (hunk 8 above)
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala]: probably not, but it doesn't matter because that Option should be changed to Optional (think I did that in other PR)

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala (hunk 3 above)
Comment (patriknw) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/ClusterSharding.scala]: > Should we avoid type rename/aliases like this in public APIs or do they look alright from Java?

IntelliJ show it as `Function` and javadoc looks good 
https://github.com/akka/akka/pull/25697#discussion_r224118919
https://doc.akka.io/japi/akka/current/akka/actor/typed/javadsl/ActorContext.html#messageAdapter-java.lang.Class-java.util.function.Function-

> Also the null bound is weird, but I think I saw another issue mentioning looking into that?

yes, that is tracked elsewhere

**Review by octonato — APPROVED** [review]
Really good. 

I think that in a near future, we must take a Lagom app and port to typed to have a good understanding of possible migration pain points. Also to be able to prove that we can consume the events from the journal.

**Comment by patriknw** [comment]
rebased

File: akka-docs/src/main/paradox/typed/cluster-sharding.md [hunk 9]
```diff
@@ -64,24 +64,36 @@ Java
 When using sharding entities can be moved to different nodes in the cluster. Persistence can be used to recover the state of
 an actor after it has moved.
 
-Taking the larger example from the @ref:[persistence documentation](persistence.md#larger-example) and making it into
-a sharded entity is the same as for a non persistent behavior. The behavior:
+Akka Persistence is based on the single-writer principle the persistent actors are typically used together
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/cluster-sharding.md]: I found this sentence hard to parse. Maybe:

Akka Persistence is based on the single-writer principle, for a particular `persitenceId` only one persistent actor instance should be active. If multiple instances were to persist events at the same time, the events would be interleaved and might not be interpreted correctly on replay. Cluster sharding is typically used together with persistence to ensure that there is only one active entity for each `persistenceId`.

**Review by chbatey — APPROVED** [review]

File: akka-docs/src/main/paradox/typed/cluster-sharding.md (hunk 9 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/cluster-sharding.md]: 👍 thanks

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/PersistentEntity.scala [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Any [[Behavior]] can be used as a sharded entity actor, but the combination of sharding and persistent
+ * actors is very common and therefore this `PersistentEntity` class is provided as convenience.
+ *
+ * It is a [[PersistentBehavior]] and is implemented in the same way. It selects the `persistenceId`
+ * automatically from the [[EntityTypeKey]] and `entityId` constructor parameters by using
+ * [[EntityTypeKey.persistenceIdFrom]].
+ */
+abstract class PersistentEntity[Command, Event, State >: Null] private (
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/PersistentEntity.scala]: good point, I renamed:
s/ShardedEntityContext/EntityContext/
s/ShardedEntity/Entity/

**Comment by patriknw** [comment]
incorporated feedback

**Comment by mvodas** [comment]
In docs (https://doc.akka.io/docs/akka/current/cluster-sharding.html#an-example):
- _"As a rule of thumb, the number of shards should be a factor ten greater than the planned maximum number of cluster nodes."_
- _"A simple sharding algorithm that works fine in most cases is to take the absolute value of the hashCode of the entity identifier modulo number of shards."_

So we would have at least one shard per node (in order to distribute our entities in the cluster, because each shard is hosted in one node, if I am not mistaken): `for (i <- 0 until cluster_size) ClusterSharding.init(Entity(EntityTypeKey[Command]("type_of_actor_" + i), ctx => behavior(ctx.entityId)))`.

Assuming that our entityId takes unique values, say we shutdown the cluster, add more nodes and start it again. The shard each persistent actor/entity should belong to has changed due to changing the number of nodes (and this is desirable in order to have a balanced distribution of actors/entities in the cluster). After this pull request, it seems the persistenceId will change as well so the actor will not find any state. So the idea is we would like to be able to decouple the persistenceId from the shards, which was possible in the previous version. With this pull request I believe this is no longer possible. Any chance for this behavior to be changed in future versions?
